### PR TITLE
Use Elemental_jll

### PR DESCRIPTION
**Pull request by christopher-dG**
See CI results: This does not work :frowning: 

Lots of segfaults happening. Not sure if they're because of the MPI version being used to run the tests, or the way that Elemental_jll was built, or what...

I wonder if https://github.com/JuliaParallel/MPI.jl/pull/367 would help, since then the MPI launching the tests would be the same as Elemental is linked against (I think).

### CODE REVIEW

**Comment by ViralBShah** [comment]
> I wonder if [JuliaParallel/MPI.jl#367](https://github.com/JuliaParallel/MPI.jl/pull/367) would help, since then the MPI launching the tests would be the same as Elemental is linked against (I think).

Yes. That will certainly help. 

cc @simonbyrne

**Comment by andreasnoack** [comment]
It's still not completely clear to me how this would work with a custom HPC-MPI. Could you please elaborate on what would happen if you try to install Elemental on a cluster with a custom MPI version?

**Comment by ViralBShah** [comment]
I think in that case we need to allow people to provide their own.

**Comment by simonbyrne** [comment]
I agree. The problem is that there isn’t really an effective mechanism for specifying which library to use. The Overrides.toml is too limited to be of much use.

**Comment by ViralBShah** [comment]
We perhaps need @staticfloat to give us a solution to this.

**Comment by christopher-dG** [comment]
I'm not really sure what's up with CI but @ViralBShah you can try out this branch if you like, I think it should work identically to master branch.

I was probably overzealous in deleting a bunch of the CI setup stuff, since we still have to use MPI.jl in the tests.

**Comment by christopher-dG** [comment]
I'll also note that despite the build failing here, it's not actually different from master -- master is setting the exit code improperly (with `&`).

**Comment by ViralBShah** [comment]
If the same tests as master are failing, how about mark them as expected to fail or disable and get this all merged?

**Comment by andreasnoack** [comment]
Let's not merge before tests pass. If there are issues with master CI then let's fix that.

**Comment by andreasnoack** [comment]
I've fixed the tests in https://github.com/JuliaParallel/Elemental.jl/pull/62 so please rebase and hopyfully tests will pass here as well.

File: Project.toml [hunk 1]
```diff
@@ -1,28 +1,30 @@
...
 Distributed = "8ba89e20-285c-5b6f-9357-94700520ee1b"
 DistributedArrays = "aaf54ef3-cdf8-58ed-94cc-d582ad619b94"
-LibGit2 = "76f85450-5226-5b5a-8eaa-529ad045b433"
+Elemental_jll = "c2e960f2-a21d-557e-aa36-859d46eed7e8"
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 
 [compat]
 DistributedArrays = "0.5, 0.6"
+Elemental_jll = "< 1"
```
Comment (andreasnoack) [inline Project.toml:14]:
```suggestion
Elemental_jll = "0.87"
```
To avoid upstream breakage

File: Project.toml [hunk 2]
```diff
@@ -1,28 +1,30 @@
...
 MPI = "0.11"
 MPIClusterManagers = "0.1"
 TSVD = "0.4"
 julia = "1"
```
Comment (andreasnoack) [inline Project.toml:18]:
```suggestion
julia = "1.3"
```
I believe it's required for the jlls

**Comment by andreasnoack** [comment]
I've no idea why we are getting this error https://travis-ci.org/github/JuliaParallel/Elemental.jl/jobs/681894268#L1211-L1225. Any ideas?

**Comment by ViralBShah** [comment]
How do I get Elemental to pick up Elemental_jll on the other Julia processes? I am using the released jll and the dev version of Elemental.jl on your branch. (edited) 
```
julia> using Elemental
[ Info: Precompiling Elemental [902c3f28-d1ec-5e7e-8399-a24c3845ee38]
┌ Warning: Package Elemental does not have Elemental_jll in its dependencies:
│ - If you have Elemental checked out for development and have
│   added Elemental_jll as a dependency but haven't updated your primary
│   environment's manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with Elemental
└ Loading Elemental_jll into Elemental from project dependency, future warnings for Elemental are suppressed.
ERROR: LoadError: UndefVarError: libEl not defined
Stacktrace:
 [1] include(::Module, ::String) at ./Base.jl:377
 [2] top-level scope at none:2
 [3] eval at ./boot.jl:331 [inlined]
 [4] eval(::Expr) at ./client.jl:449
 [5] top-level scope at ./none:3
in expression starting at /home/viralbshah/.julia/dev/Elemental/src/Elemental.jl:10
ERROR: Failed to precompile Elemental [902c3f28-d1ec-5e7e-8399-a24c3845ee38] to /home/viralbshah/.julia/compiled/v1.4/Elemental/mJwGy_jmS8y.ji.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at ./loading.jl:1272
 [3] _require(::Base.PkgId) at ./loading.jl:1029
 [4] require(::Base.PkgId) at ./loading.jl:927
 [5] require(::Module, ::Symbol) at ./loading.jl:922
```

**Comment by giordano** [comment]
Did you run `] resolve`?

**Comment by ViralBShah** [comment]
As in should I do `@everywhere Pkg.resolve()`? I have it working reliably on a single process. It is the parallel that I am having trouble with.

**Comment by giordano** [comment]
Uh, sorry, misread the question. That's strange :confused:

**Comment by christopher-dG** [comment]
There's also still this MPI build error [here](https://travis-ci.org/github/JuliaParallel/Elemental.jl/jobs/681935230#L408-L410) that seems like a bug in the MPI build script.

**Comment by christopher-dG** [comment]
I wonder if the other Julia processes are only using the package UUID, and so are getting the current released (without JLL) version of Elemental? Is the manifest used here? I'm not really sure.

**Comment by christopher-dG** [comment]
Do we want to keep the `extras/elpp.jl` stuff? It should just need a tweak to update the libEl installation path.

**Comment by ViralBShah** [comment]
Pinging @simonbyrne in here (since he may be interested to follow the progress and issues in here.)

**Comment by simonbyrne** [comment]
Hmm, not sure what the problem is. I'll try to spend some time finishing up https://github.com/JuliaParallel/MPI.jl/pull/367 next week.

**Comment by christopher-dG** [comment]
The Mac issues are fixed, but I don't know what to make of this:

```
ERROR: LoadError: IOError: mkdir: file already exists (EEXIST)
Stacktrace:
 [1] uv_error at ./libuv.jl:97 [inlined]
 [2] mkdir(::String; mode::UInt16) at ./file.jl:177
 [3] mkpath(::String; mode::UInt16) at ./file.jl:227
 [4] mkpath at ./file.jl:222 [inlined]
 [5] compilecache_path(::Base.PkgId) at ./loading.jl:1210
 [6] compilecache(::Base.PkgId, ::String) at ./loading.jl:1240
 [7] _require(::Base.PkgId) at ./loading.jl:1029
 [8] require(::Base.PkgId) at ./loading.jl:927
 [9] require(::Module, ::Symbol) at ./loading.jl:922
 [10] include(::Module, ::String) at ./Base.jl:377
 [11] exec_options(::Base.JLOptions) at ./client.jl:288
 [12] _start() at ./client.jl:484
in expression starting at /Users/travis/build/JuliaParallel/Elemental.jl/test/tsvd.jl:1
```

**Comment by simonbyrne** [comment]
Potentially a race condition in precompilation?

**Comment by christopher-dG** [comment]
Hm, that's a good point -- it's importing a test-only dependency so the `precompile` in the `before_script` won't catch it.

**Comment by andreasnoack** [comment]
We could also just disable precompilation in CI since it's also probably slowing down the run.

**Comment by christopher-dG** [comment]
Looks like that part is working now at least. It's trying to run a bare `mpiexec` now, which is confusing because I think MPI and MPIClusterManagers should be picking up the MPICH_jll one now.

**Comment by ViralBShah** [comment]
In order to merge this, do we essentially need to fix:

https://github.com/JuliaParallel/MPIClusterManagers.jl/issues/16

**Comment by christopher-dG** [comment]
In order for tests to pass, I think so.

**Comment by christopher-dG** [comment]
There still seems to be some race conditions in CI sometimes, but I think we can probably squash + merge this now?

**Comment by ViralBShah** [comment]
Go for it!

**Pull request merged**